synapse/storage/databases/main/user_directory.py

```diff
@@ -369,6 +371,22 @@ def _get_next_batch(txn: LoggingTransaction) -> Optional[List[str]]:
 
         return len(users_to_work_on)
 
+    async def is_excluded_from_user_dir(self, user: str) -> bool:
+        """Certain classes of local user are omitted from the user directory.
+        Is this user one of them?
+        """
+        if self.hs.is_mine_id(user):
+            # Support users are for diagnostics and should not appear in the user directory.
+            if await self.is_support_user(user):
+                return True
+            # Deactivated users aren't contactable, so should not appear in the user directory.
+            if await self.get_user_deactivated_status(user):
+                return True
+            # App service users aren't usually contactable, so exclude them.
+            if self.get_if_app_services_interested_in_user(user):
+                return True
```

Comment:
I'm unsure what to think about this.
I have some appservice-generated users in *my* homeserver's search directory. I can't confirm or deny whether I've done something wrong to wind up with that happening, but thought it was worth mentioning in case this should trigger the question of 'do we really intend to exclude AS users from the directory?'.

Both yes and no make sense here. No, on a lot of networks (IRC probably? Discord when using a guild-bot.), the users aren't contactable.
But yes, on some networks, users are contactable and there's nothing inherently wrong with letting you discover them when trying to start a DM ...

I would guess that the 'No' cases outweigh the 'Yes' cases, so I think what you're doing is sensible.

Reply:
It seems like *most* of the code currently excludes AS users from the directory, but there are enough bugs that they leak through.

At least we're making it consistent here. There might be scope for a future change where we make this configurable per-appservice, but let's leave it for now.